README.md
# AnalogIO library for Arduino
## Description
This library allows to read a value from an analog input like an potentiometer, or from a digital input like an encoder. Moreover, allows to write it on digital output, exactly on PWM pin. You can specify the device, and the resolution of reading and writing. If you want, there is a controller class that manages an object AnalogIn and/or an object AnalogOut, without worrying about the normalization if the resolutions of the objects are different.

## Manual installation
To install manually this library, you can see this little guides.

### Arduino IDE
1. Download and extract it
2. Open Arudino IDE
3. Select tab `Sketch` -> `Include Library` -> `Add .ZIP Library`
4. Select the zip file included in `extras`
You can select the library on `Include Library` -> `AnalogIO`. It will be automatically added on you project, like this:
```c++
#include <AnalogIO.h>
```

### Other IDE
1. Download and extract it
2. Copy the `src` content in your project folder
3. Open the project with your IDE
4. Use the preprocessor directives to add the library, like this:
```c++
#include "AnalogIO.h"
```
## How to use
See the [Wiki](https://github.com/davidepalladino/AnalogIO-Arduino/wiki) page for details about the use.

## Changelog
See the [Changelog](/CHANGELOG.md) page for details about the versions.

## Contribute
Your contribute is important for me. Don't hesitate with issues and pull requests for improving this class.

### Special thanks
Special thanks to my cousin [Marco Palladino](https://github.com/PalladinoMarco) for the help to optimize the code.


CHANGELOG.md
In this document there is the story about the various versions. See the [Wiki](https://github.com/davidepalladino/AnalogIO-Arduino/wiki) page for detail about the use.

### 1.1.1 - 2020-07-06
**Added**
* Now you can get the max value of the object `AnalogIn` from the object `AnalogController`.
* Now you can get the max value of the object `AnalogOut` from the object `AnalogController`.

**Changed**
* Value of `VALUE_NONE` from `0` to the max value of 32 bit.
* Return type of `getMaxValue`, for `AnalogIn`, from `uint16_t` to `int32_t`.
* Parameter type of `timeOut`, for `AnalogController`, from `uint8_t` to `uint32_t`.

**Fixed**
* Bug related to emulation of analog pins.

src/AnalogIO.h
 /**
  * This library allows to read a value from an analog input like an potentiometer, or from a digital input like an encoder. Moreover, allows to write it on digital output, exactly on PWM pin. 
  * You can specify the device, and the resolution of reading and writing. If you want, there is a controller class that manages an object "AnalogIn" and/or an object "AnalogOut", 
  *  without worrying about the normalization if the resolutions of the objects are different.
  * If you want more information about the classes, you can read the respective file header or the Wiki page on GitHub.
  * Copyright (c) 2020 Davide Palladino. 
  * All right reserved.
  * 
  * @author Davide Palladino
  * @contact davidepalladino@hotmail.com
  * @version 1.0
  * @date June 21th, 2020
  * 
  * This library is free software; you can redistribute it and/or
  *  modify it under the terms of the GNU General Public
  *  License as published by the Free Software Foundation; either
  *  version 3.0 of the License, or (at your option) any later version
  * 
  * This library is distributed in the hope that it will be useful,
  *  but WITHOUT ANY WARRANTY; without even the implied warranty of
  *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the 
  *  GNU Lesser General Public License for more details.
  * 
  */

#include "incl/AnalogController.h"
#include "incl/AnalogIn.h"
#include "incl/AnalogOut.h"

src/incl/AnalogIn.h
/**
  * This library allows to read a value from an analog input like an potentiometer, or from a digital input like an encoder.
  * You can specify the device and the resolution of reading.
  * Copyright (c) 2020 Davide Palladino. 
  * All right reserved.
  * 
  * @author Davide Palladino
  * @contact davidepalladino@hotmail.com
  * @version 1.1.1
  * @date 6th July, 2020
  * 
  * This library is free software; you can redistribute it and/or
  *  modify it under the terms of the GNU General Public
  *  License as published by the Free Software Foundation; either
  *  version 3.0 of the License, or (at your option) any later version
  * 
  * This library is distributed in the hope that it will be useful,
  *  but WITHOUT ANY WARRANTY; without even the implied warranty of
  *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the 
  *  GNU Lesser General Public License for more details.
  * 
  */

#ifndef ANALOGIN_H
    #define ANALOGIN_H

    #include "globalData.h"

    /**
     * Class for initialize and manage the analog/digital input.
     */
    class AnalogIn {      
        /* Allow the AnalogController class to access a private methods and attributes. */
        friend class AnalogController;   

        public:
            /** 
             * This constructor creates the object setting only the pin, if the device is a potentiometer.
             * @param pin Analog pin where the value will be read, if the device is a potentiometer.
             * @warning The resolution of reading will be 10 bit.
             */
            AnalogIn(uint8_t pin);

            /** 
             * This constructor creates the object setting the pin and the resolution of reading, if the device is a potentiometer.
             * @param pin Analog pin where the value will be read, if the device is a potentiometer.
             * @param resolution Resolution of reading between "BIT4", "BIT6", "BIT8", "BIT10" and "BIT12".
             * @warning Resolution "BIT12" is only for boards DUE, ZERO or MKR family. If you try to set it with a different board, the resolution will be "BIT10".
             *  "NONE" is for encoder's device; if you try to set it, the resolution will be "BIT10."
             */
            AnalogIn(uint8_t pin, resolution_t resolution);

            /** 
             * This constructor creates the object setting only the pins, if the device is an ancoder. Moreover, will be called "pinMode" in "INPUT" mode.
             * @param pinA First digital pin where the value will be read.
             * @param pinB Second digital pin where the value will be read.
             * @warning The resolution of reading will be 10 bit.
             */
            AnalogIn(uint8_t pinA, uint8_t pinB);

            /** 
             * This constructor creates the object setting the pins and the resolution of reading, if the device is an ancoder. Moreover, will be called "pinMode" in "INPUT" mode.
             * @param pinA First digital pin where the value will be read.
             * @param pinB Second digital pin where the value will be read.
             * @param resolution Resolution of reading between "NONE", "BIT4", "BIT6", "BIT8", "BIT10" and "BIT12". 
             *  If you set to "NONE", the min and max value will be that the possible at 32 bit.
             */
            AnalogIn(uint8_t pinA, uint8_t pinB, resolution_t resolution);

            /**
             * This method sets the speed of variation of the value, if the device is an encoder.
             * @param speed Value between "MIN_SPEED" and "MAX_SPEED".
             */
            void setSpeed(uint8_t speed);

            /**
             * This method gets the max value that can be read. 
             * @return Max value that can be read, or the max possible at 32 bit if the resolution is set to "NONE".
             */
            int32_t getMaxValue();

            /**
             * This method gets the speed of variation of the value, if the device is an encoder.
             * @return Value between "MIN_SPEED" and "MAX_SPEED"; 0 if the device is a potentiometer.
             */
            uint8_t getSpeed();

            /**
             * This method reads the value from a potentiometer.
             * @return Value between "MIN_VALUE_BIT" and "maxValue".
             */     
            uint16_t read();

             /**
             * This method reads the value from an encoder. Moreover, there will be the debouncing.
             * @param value Original value to start.
             * @return Value between "MIN_VALUE_BIT" and "maxValue", incremented/decremented according to the speed value. If the resolution is set to "NONE", the min and max value will be that the possible at 32 bit.
             */     
            int32_t read(int32_t value);

        private:
            uint8_t pinA;                       // Pin where the value will be read.
            uint8_t pinB;                       // Pin where the value will be read if the device is an encoder.
            device_t device;                    // Type of device, between "POTENTIOMETER" and "ENCODER".
            resolution_t resolution;            // Type of resolution.
            int32_t maxValue;                   // Max value that can be read.
            uint8_t speed;                      // Speed of variation of the value, if the device is an encoder.
            uint8_t lastStatusEncoder;          // Last status of the encoder, to allow the debouncing.
            bool emulatePinA;                   // Flag to indicate if the "pinA" will be emulated or not.
            bool emulatePinB;                   // Flag to indicate if the "pinB" will be emulated or not.

            /**
             * This method sets the pin of the device.
             * @param pinA Analog pin where the value will be read, if the device is a potentiometer; else, digital pin if the device is an encoder.
             */
            void setPinA(uint8_t pinA);

            /**
             * This method sets the pin of the device if is encoder type.
             * @param pinB Digital pin where the value will be read, if the device is an encoder type.
             */
            void setPinB(uint8_t pinB);

            /**
             * This method sets the type of the device.
             * @param device Value between "POTENTIOMETER" and "ENCODER".
             */
            void setDevice(device_t device);

            /**
             * This method sets the type of the resolution.
             * @param resolution Value between "NONE", "BIT4", "BIT6", "BIT8", "BIT10" and "BIT12".
             */
            void setResolution(resolution_t resolution);

            /**
             * This method sets the max value that can be read.
             * @param maxValue Max value that can be read.
             */
            void setMaxValue(int32_t maxValue);

            /**
             * This method sets the last status of the encoder.
             * @param lastStatusEncoder Value between 0 and 1.
             */                   
            void setLastStatusEncoder(uint8_t lastStatusEncoder);

            /**
             * This method sets if the "pinA" will be emulated or not in digital.
             * @param status Boolean value "true" if the pin will be emulated; "false" if will not be.
             */
            void setEmulatePinA(bool status);

            /**
             * This method sets if the "pinB" will be emulated or not in digital.
             * @param status Boolean value "true" if the pin will be emulated; "false" if will not be.
             */
            void setEmulatePinB(bool status);

            /**
             * This method gets the "pinA".
             * @return Analog pin if the device is a potentiometer; else, digital pin if the device is an encoder.
             */         
            uint8_t getPinA();

            /**
             * This method gets the "pinB".
             * @return Digital pin if the device is an encoder.
             */        
            uint8_t getPinB();

            /**
             * This method gets the type of the device.
             * @return Value between "POTENTIOMETER" and "ENCODER".
             */    
            device_t getDevice();

            /**
             * This method gets the type of the resolution.
             * @return Value between "NONE", "BIT4", "BIT6", "BIT8", "BIT10" and "BIT12".
             */    
            resolution_t getResolution();

            /**
             * This method gets the last status of the encoder.
             * @return Value between 0 and 1.
             */  
            uint8_t getLastStatusEncoder();

            /**
             * This method gets if the "pinA" will be emulated or not in digital.
             * @return Boolean value "true" if the pin will be emulated; "false" if will not be.
             */ 
            bool getEmulatePinA();

            /**
             * This method gets if the "pinB" will be emulated or not in digital.
             * @return Boolean value "true" if the pin will be emulated; "false" if will not be.
             */ 
            bool getEmulatePinB(); 
    };
#endif

src/incl/AnalogOut.h
/**
  * This library allows to write a value on digital output, exactly on PWM pin. 
  * You can specify the resolution of writing.
  * Copyright (c) 2020 Davide Palladino. 
  * All right reserved.
  * 
  * @author Davide Palladino
  * @contact davidepalladino@hotmail.com
  * @version 1.0
  * @date 21th June, 2020
  * 
  * This library is free software; you can redistribute it and/or
  *  modify it under the terms of the GNU General Public
  *  License as published by the Free Software Foundation; either
  *  version 3.0 of the License, or (at your option) any later version
  * 
  * This library is distributed in the hope that it will be useful,
  *  but WITHOUT ANY WARRANTY; without even the implied warranty of
  *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the 
  *  GNU Lesser General Public License for more details.
  * 
  */

#ifndef ANALOGOUT_H
    #define ANALOGOUT_H
    
    #include "globalData.h"

    /**
     * Class for initialize and manage the analog output.
     */
    class AnalogOut {
        /* Allow the AnalogController class to access a private methods and attributes. */
        friend class AnalogController;

        public:
            /** 
             * This constructor creates the object setting only the pin of writing. 
             * @param pin Digital pin where the value will be written.
             * @warning The resolution of writing will be "BIT8".
             */
            AnalogOut(uint8_t pin);

            /** 
             * This constructor creates the object setting the pin and the resolution of writing. 
             * @param pin Digital pin where the value will be written.
             * @param resolution Resolution of writing between "BIT4", "BIT6", "BIT8", "BIT10" and "BIT12".
             * @warning The resolutions "BIT10" and "BIT12" is only for boards DUE, ZERO or MKR family. If you try to set it with a different board, the resolution will be "BIT8".
             */
            AnalogOut(uint8_t pin, resolution_t);

            /**
             * This method gets the max value that can be written.
             * @return Max value that can be written.
             */
            uint16_t getMaxValue();

            /**
             * This method writes the value to the pin.
             * @param value Value to write, between "MIN_VALUE_BIT" and "maxValue".
             * @return Value written.
             */     
            uint16_t write(uint16_t value);

        private:
            uint8_t pin;                        // Pin where the value will be written. 
            resolution_t resolution;            // Type of resolution.
            uint16_t maxValue;                  // Max value that can be written.

            /**
             * This method sets the pin where the value will be written.
             * @param pin Digital pin where the value will be written.
             */
            void setPin(uint8_t pin);

            /**
             * This method sets the type of the resolution.
             * @param resolution Value between "BIT8", "BIT10" and "BIT12".
             */
            void setResolution(resolution_t resolution);

            /**
             * This method sets the max value that can be written.
             * @param maxValue Max value that can be written.
             */
            void setMaxValue(uint16_t maxValue);

            /**
             * This method gets the pin where the value will be written. 
             * @return Digital pin where the value will be written. 
             */         
            uint8_t getPin();

            /**
             * This method gets the type of the resolution.
             * @return Value between "BIT8", "BIT10" and "BIT12".
             */    
            resolution_t getResolution();
    };
#endif

src/incl/AnalogIn.cpp
#include "AnalogIn.h"

AnalogIn::AnalogIn(uint8_t pin) : AnalogIn (pin, BIT10) {}

AnalogIn::AnalogIn(uint8_t pin, resolution_t resolution) {
    setPinA(pin);

    setDevice(POTENTIOMETER);

    setResolution(resolution);

    switch (getResolution()) {
        case BIT4:
            setMaxValue(MAX_VALUE_BIT4);
            break;
        case BIT6:
            setMaxValue(MAX_VALUE_BIT6);
            break;
        case BIT8:
            setMaxValue(MAX_VALUE_BIT8);
            break;
        case BIT10:
        case NONE:
            setMaxValue(MAX_VALUE_BIT10);
            break;
        case BIT12:
            /* Check if the board is DUE, ZERO or MKR family. */
            #if defined(ARDUINO_ARCH_SAM) || defined(ARDUINO_ARCH_SAMD)
                analogReadResolution((uint8_t) getResolution());
                setMaxValue(MAX_VALUE_BIT12);
            #else
                setResolution(BIT10);
                setMaxValue(MAX_VALUE_BIT10);
            #endif
            break;
    }
}

AnalogIn::AnalogIn(uint8_t pinA, uint8_t pinB) : AnalogIn(pinA, pinB, BIT10) {}

AnalogIn::AnalogIn(uint8_t pinA, uint8_t pinB, resolution_t resolution) {
    setPinA(pinA);
    setPinB(pinB);
    pinMode(getPinA(), INPUT);
    pinMode(getPinB(), INPUT);

    setDevice(ENCODER);

    setResolution(resolution);

    switch (getResolution()) {
        case NONE:
            setMaxValue(VALUE_NONE);
            break;
        case BIT4:
            setMaxValue(MAX_VALUE_BIT4);
            break;
        case BIT6:
            setMaxValue(MAX_VALUE_BIT6);
            break;
        case BIT8:
            setMaxValue(MAX_VALUE_BIT8);
            break;
        case BIT10:
            setMaxValue(MAX_VALUE_BIT10);
            break;
        case BIT12:
            setMaxValue(MAX_VALUE_BIT12);
            break;
    }

    setSpeed(MIN_SPEED);

    setLastStatusEncoder(HIGH);

    setEmulatePinA((bool) !digitalRead(getPinA()));
    setEmulatePinB((bool) !digitalRead(getPinB()));
}

void AnalogIn::setSpeed(uint8_t speed) {
    /* Check if the device is an encoder. In this case, will be set the speed. */
    if (getDevice() == ENCODER) {
        if (speed < MIN_SPEED) {
            this->speed = MIN_SPEED;
        } else if (speed > MAX_SPEED) {
            this->speed = MAX_SPEED;
        } else {
            this->speed = speed;
        }
    }
}

int32_t AnalogIn::getMaxValue() { return this->maxValue; }

uint8_t AnalogIn::getSpeed() { 
    /* Check if the device is an encoder. In this case, will return the speed; else will return 0. */
    if (getDevice() == ENCODER) {
        return this->speed; 
    } else {
        return 0;
    }
}

uint16_t AnalogIn::read() {
    /* Check if the device is a potentiometer. In this case, will be read the value; else will return 0. */
    if (getDevice() == POTENTIOMETER) {
        if (getResolution() < BIT10) {
            return map(analogRead(getPinA()), MIN_VALUE_BIT, MAX_VALUE_BIT10, MIN_VALUE_BIT, getMaxValue());
        } else {
            return analogRead(getPinA());
        }
    } else {
        return 0;
    }
}

int32_t AnalogIn::read(int32_t value) {
    /* Check if the device is an encoder. In this case, will be read the value; else will return 0. */
    if (getDevice() == ENCODER) {
        /* Create a variable to save the last status of the encoder. */
        uint8_t currentStatusEncoder;

        /* Proceed with the first reading of the encoder, checking if the pin emulates or not the digital pin. */
        if (getEmulatePinA()) {
            currentStatusEncoder = analogRead(getPinA()) > 100 ? 1 : 0;
        } else {
            currentStatusEncoder = digitalRead(getPinA());
        }

        /* Proceed with the second reading of the encoder, checking if the pin emulates or not the digital pin. */
        if ((getLastStatusEncoder() == HIGH) && (currentStatusEncoder == LOW)) {
            if (getEmulatePinB()) {
                if (analogRead(getPinB()) > 100 ? 1 : 0) {
                    value += getSpeed();
                } else {
                    value -= getSpeed();
                }
            } else {
                if (digitalRead(getPinB()) == HIGH) {
                    value += getSpeed();
                } else {
                    value -= getSpeed();
                }
            }
        }

        /* If the resolution is not "NONE", the value will be correct if is greater than "maxValue" or less than "MIN_VALUE_BIT" */
        if (getResolution() != NONE) {
            if (value < MIN_VALUE_BIT) {
                value = MIN_VALUE_BIT;
            } else if (value >= getMaxValue()) {
                value = getMaxValue();
            }
        }
       
        /* Save the last status of the encoder with the current, for next reading. */
        setLastStatusEncoder(currentStatusEncoder);

        return value;
    } else {
        return 0;
    }
}

void AnalogIn::setPinA(uint8_t pinA) { this->pinA = pinA; }

void AnalogIn::setPinB(uint8_t pinB) { this->pinB = pinB; }

void AnalogIn::setDevice(device_t device) { this->device = device; }

void AnalogIn::setResolution(resolution_t resolution) { this->resolution = resolution; }

void AnalogIn::setMaxValue(int32_t maxValue) { this-> maxValue = maxValue; }

void AnalogIn::setLastStatusEncoder(uint8_t lastStatusEncoder) { this->lastStatusEncoder = lastStatusEncoder; }

void AnalogIn::setEmulatePinA(bool status) { this->emulatePinA = status; }

void AnalogIn::setEmulatePinB(bool status) { this->emulatePinB = status; }

uint8_t AnalogIn::getPinA() { return this->pinA; }

uint8_t AnalogIn::getPinB() { return this->pinB; }

device_t AnalogIn::getDevice() { return this->device; }

resolution_t AnalogIn::getResolution() { return this->resolution; }

uint8_t AnalogIn::getLastStatusEncoder() { return this->lastStatusEncoder; }

bool AnalogIn::getEmulatePinA() { return this->emulatePinA; }

bool AnalogIn::getEmulatePinB() { return this->emulatePinB; }

src/incl/AnalogOut.cpp
#include "AnalogOut.h"

AnalogOut::AnalogOut(uint8_t pin) : AnalogOut(pin, BIT8) {}

AnalogOut::AnalogOut(uint8_t pin, resolution_t resolution) {
    setPin(pin);
    pinMode(getPin(), OUTPUT);

    setResolution(resolution);

    switch (getResolution()) {
        case BIT4:
            setMaxValue(MAX_VALUE_BIT4);
            break;
        case BIT6:
            setMaxValue(MAX_VALUE_BIT6);
            break;
        case BIT8:
        case NONE:
            setMaxValue(MAX_VALUE_BIT8);
            break;
        case BIT10:
            /* Check if the board is DUE, ZERO or MKR family. */
            #if defined(ARDUINO_ARCH_SAM) || defined(ARDUINO_ARCH_SAMD)
                analogWriteResolution((uint8_t) getResolution());
                setMaxValue(MAX_VALUE_BIT10);
            #else
                setResolution(BIT8);
                setMaxValue(MAX_VALUE_BIT8);
            #endif
            break;
        case BIT12:
            /* Check if the board is DUE, ZERO or MKR family. */
            #if defined(ARDUINO_ARCH_SAM) || defined(ARDUINO_ARCH_SAMD)
                analogWriteResolution((uint8_t) getResolution());
                setMaxValue(MAX_VALUE_BIT12);
            #else
                setResolution(BIT8);
                setMaxValue(MAX_VALUE_BIT8);
            #endif
            break;
    }
}

uint16_t AnalogOut::getMaxValue() { return this->maxValue; }

uint16_t AnalogOut::write(uint16_t value) {
    /* Check if the value is greater than "maxValue", by setting it if so. */
    if (value > getMaxValue()) {
        value = getMaxValue();
    }

    /* Normalize the value if the resolution is less than "BIT8". */
    if (getResolution() < BIT8) {
        value = map(value, MIN_VALUE_BIT, getMaxValue(), MIN_VALUE_BIT, MAX_VALUE_BIT8);
    }
    
    analogWrite(getPin(), value);

    return value;
}

void AnalogOut::setPin(uint8_t pin) { this->pin = pin; }

void AnalogOut::setResolution(resolution_t resolution) { this->resolution = resolution; }

void AnalogOut::setMaxValue(uint16_t maxValue) { this-> maxValue = maxValue; }

uint8_t AnalogOut::getPin() { return this->pin; }

resolution_t AnalogOut::getResolution() { return this->resolution; }

src/incl/AnalogController.cpp
#include "AnalogController.h"

AnalogController::AnalogController(AnalogIn* objIn, AnalogOut* objOut){
    setObjIn(objIn);
    setObjOut(objOut);

    setValue(MIN_VALUE_BIT);
}

void AnalogController::setValue(int32_t value) {
    /* Check if the "objIn" exists, and in this case will be considered the "MIN_VALUE_BIT" and the "maxValue" if the resolution is not equal to "NONE". */
    if (getObjIn() != NULL) {
        if (getObjIn()->getResolution() != NONE) {
            if (value < MIN_VALUE_BIT) {
                this->value = MIN_VALUE_BIT;
            } else if (value > getMaxValueIn()) {
                this->value = getMaxValueIn();
            } else {
                this->value = value;
            }
        } else {
            this->value = value;
        }
    /* If the "objIn" doesn't exist, will be considered the "objOut". */
    } else if ((getObjIn() == NULL) && (getObjOut() != NULL)) {
        if (value < MIN_VALUE_BIT) {
            this->value = MIN_VALUE_BIT;
        } else if (value > getMaxValueOut()) {
            this->value = getMaxValueOut();
        } else {
            this->value = value;
        }
    }
}

void AnalogController::setSpeed(uint8_t speed) {
    if (getObjIn() != NULL) {
        getObjIn()->setSpeed(speed);
    }
}

int32_t AnalogController::getValue() { return this->value; }

void AnalogController::readValue() {
    if (getObjIn() != NULL) {
        /* Check the device, to execute the correct method. */
        if (getObjIn()->getDevice() == POTENTIOMETER) {
            setValue(getObjIn()->read());
        } else if (getObjIn()->getDevice() == ENCODER) {
            setValue(getObjIn()->read(getValue()));
        }
    }
}


int32_t AnalogController::getMaxValueIn() {
    if (getObjIn() != NULL) {
        return getObjIn()->getMaxValue();
    } else {
        return 0;
    }
}

uint16_t AnalogController::getMaxValueOut() {
    if (getObjOut() != NULL) {
        return getObjOut()->getMaxValue();
    } else {
        return 0;
    }
}

uint8_t AnalogController::getSpeed() {
    if (getObjIn() != NULL) {
        return getObjIn()->getSpeed();
    } else {
        return 0;
    }
}

uint16_t AnalogController::writeValue() {
    return writeValue(100);
}

uint16_t AnalogController::writeValue(double percent) {
    if (getObjOut() != NULL) {
        int32_t valueTemp = getValue();

        /* Normalize the value if the resolution is not similar between "objIn" and "objOut". */
        if ((getObjIn() != NULL) && (getObjIn()->getResolution() != NONE) && (getObjIn()->getResolution() != getObjOut()->getResolution())) {
            valueTemp = map((uint16_t) valueTemp, MIN_VALUE_BIT, getMaxValueIn(), MIN_VALUE_BIT, getMaxValueOut());
        /* Force the value to "MIN_VALUE_BIT" or "maxValue" of "objOut" if the value is greater or less than them, and the resolution of "objIn" is set to "NONE". */
        } else if ((getObjIn() != NULL) && (getObjIn()->getResolution() == NONE)) {
            if (valueTemp < MIN_VALUE_BIT) {
                valueTemp = MIN_VALUE_BIT;
            } else if (valueTemp > getMaxValueOut()) {
                valueTemp = getMaxValueOut();
            }
        }
        
        /* Calculate the percent, if is less than 100. */
        if (percent != 100) {
            valueTemp = ((double) valueTemp * percent) / 100;
        }

        return getObjOut()->write((uint16_t) valueTemp);
    } else {
        return 0;
    }
}

void AnalogController::blink(uint8_t times, uint32_t timeOut) {
    if (getObjOut() != NULL) {
        for (uint8_t t = 1; t <= times; t++) {
            getObjOut()->write(getMaxValueOut());
            delay(timeOut);
            getObjOut()->write(MIN_VALUE_BIT);

            if(t < times) {
                delay(timeOut);
            }
        }    
    }
}

void AnalogController::setObjIn(AnalogIn* objIn) {
    this->objIn = objIn;
}

void AnalogController::setObjOut(AnalogOut* objOut) {
    this->objOut = objOut;
}

AnalogIn* AnalogController::getObjIn() {
    return this->objIn;
}

AnalogOut* AnalogController::getObjOut() {
    return this->objOut;
}

src/incl/AnalogController.h
 /**
  * This library allows to manage an object "AnalogIn" and/or an object "AnalogOut". Indeed you can read, write, manually set or get the value, without worrying about the normalization 
  *  if the resolutions of the objects are different. Every method has the right controls. 
  * If you want to use this class, initially you will have to create the rispective object "AnalogIn" and/or "AnalogOut", for passing them to the constructor. For more information about this class,
  *  read the descriptions of the methods.
  * Copyright (c) 2020 Davide Palladino. 
  * All right reserved.
  * 
  * @author Davide Palladino
  * @contact davidepalladino@hotmail.com
  * @version 1.1.1
  * @date 6th July, 2020
  * 
  * This library is free software; you can redistribute it and/or
  *  modify it under the terms of the GNU General Public
  *  License as published by the Free Software Foundation; either
  *  version 3.0 of the License, or (at your option) any later version
  * 
  * This library is distributed in the hope that it will be useful,
  *  but WITHOUT ANY WARRANTY; without even the implied warranty of
  *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the 
  *  GNU Lesser General Public License for more details.
  * 
  */

#ifndef ANALOGCONTROLLER_H
    #define ANALOGCONTROLLER_H

    #include "AnalogIn.h"
    #include "AnalogOut.h"

    class AnalogController {
        public:
            /** 
             * This constructor creates the main object, setting the pointers to the objects where the value will be read and/or write.
             * @param objIn Pointer to object where the value will be read. Set it to "NULL" if will be not used.
             * @param objOut Pointer to object where the value will be written. Set it to "NULL" if will be not used.
             */
            AnalogController(AnalogIn* objIn, AnalogOut* objOut);

            /**
             * This method sets the value. The max value will be considered from "objIn" if exists, and its resolution is not equal to "NONE"; else, the max value will be considered from "objOut".
             *  In example, if you set a resolution of "BIT8" on one of the objects, you can externally set a value between 0 and 255. A value less than 0 will be set to this; greater than 255 will be set to this. Else, if the resolution
             *  is set to "NONE", you can set any value at 32 bit.
             * @param value Value to store.
             */
            void setValue(int32_t value);

            /**
             * This method sets the speed of variation of the value, if the device is an encoder.
             * @param speed Value between "MIN_SPEED" and "MAX_SPEED".
             */
            void setSpeed(uint8_t speed);

            /**
             * This method gets the value.
             * @return Value stored.
             */
            int32_t getValue();

            /**
             * This method gets the max value of "objIn".
             * @return Max value of "objIn" or 0 if "objIn" doesn't exist.
             */
            int32_t getMaxValueIn();

            /**
             * This method gets the max value of "objOut".
             * @return Max value of "objOut" or 0 if "objOut" doesn't exist.
             */
            uint16_t getMaxValueOut();

            /**
             * This method gets the speed of variation of the value, if the device is an encoder.
             * @return Value between "MIN_SPEED" and "MAX_SPEED"; 0 if the device is a potentiometer or "objIn" doesn't exist.
             */
            uint8_t getSpeed();

            /**
             * This method reads and stores the value from a device, with the "objIn".
             */
            void readValue();

            /**
             * This method writes the value stored to a device, with the "objOut".
             * @return The value that was written, or 0 if "ogjOut" doesn't exist.
             * @warning If the value is less than "MIN_VALUE_BIT" or greater of "maxValue" of "objOut", will be normalized.
             */
            uint16_t writeValue();

            /**
             * This method writes a percent of the value stored to a device, with the "objOut".
             * @param percent Percent that will be written to the pin.
             * @return The value that was written, considering the percent. 
             * @warning If the value is less than "MIN_VALUE_BIT" or greater of "maxValue" of "objOut", will be normalized.
             */
            uint16_t writeValue(double percent);

            /**
             * This method activates the blink for "times" in "timeout" milliseconds, with the "objOut". 
             * @param times The number of times.
             * @param timeOut The time from blink to another.
             */
            void blink(uint8_t times, uint32_t timeOut);


        private:
            AnalogIn* objIn;               // Pointer to object where the value will be read. This parameter will be "NULL" if is not set.
            AnalogOut* objOut;             // Pointer to object where the value will be written. This parameter will be "NULL" if is not set.
            int32_t value;                 // Current value read, write or set externally.

            /**
             * This method sets the pointer to object where the value will be read.
             * @param objIn Pointer to object where the value will be read.
             */
            void setObjIn(AnalogIn* objIn);

            /**
             * This method sets the pointer to object where the value will be written.
             * @param objOut Pointer to object where the value will be written.
             */
            void setObjOut(AnalogOut* objOut);

            /**
             * This method gets the pointer to object where the value will be read.
             * @return Pointer to object.
             */
            AnalogIn* getObjIn();

            /**
             * This method gets the pointer to object where the value will be written.
             * @return Pointer to object.
             */
            AnalogOut* getObjOut();            
    };
#endif